onnx/defs/math/defs.cc

```diff
@@ -3392,4 +3392,676 @@ ONNX_OPERATOR_SET_SCHEMA(
+                "If onesided is 0, the following shape is expected: [batch_idx][signal_dim1][signal_dim2]...[signal_dimN][2]. "
+                "If axis=0 and onesided is 1, the following shape is expected: [batch_idx][floor(signal_dim1/2)+1][signal_dim2]...[signal_dimN][2]. "
+                "If axis=1 and onesided is 1, the following shape is expected: [batch_idx][signal_dim1][floor(signal_dim2/2)+1]...[signal_dimN][2]. "
+                "If axis=N-1 and onesided is 1, the following shape is expected: [batch_idx][signal_dim1][signal_dim2]...[floor(signal_dimN/2)+1][2]. "
+                "The signal_dim at the specified axis is equal to the dft_length.",
+                "T1")
+            .TypeConstraint(
+                "T1",
+                {"tensor(float16)", "tensor(float)", "tensor(double)", "tensor(bfloat16)"},
+                "Constrain input and output types to float tensors.")
+            .TypeConstraint(
+                "T2",
+                {"tensor(int64)"},
+                "Constrain scalar length types to int64_t.")
+            .TypeAndShapeInferenceFunction([](ONNX_NAMESPACE::InferenceContext& ctx) {
+                bool is_onesided = static_cast<bool>(getAttribute(ctx, "onesided", 0));
+                bool inverse = static_cast<bool>(getAttribute(ctx, "inverse", 0));
+
+                if (inverse && is_onesided) {
+                    fail_shape_inference("is_onesided and inverse attributes cannot be enabled at the same time");
+                }
+
+                propagateElemTypeFromInputToOutput(ctx, 0, 0);
+                if (!hasInputShape(ctx, 0))
+                {
+                    // If no shape is available for the input, skip shape inference...
+                    return;
+                }
+
+                // In general the output shape will match the input shape exactly
+                // So initialize the output shape with the input shape
+                auto& input_shape = getInputShape(ctx, 0);
+                ONNX_NAMESPACE::TensorShapeProto result_shape_proto = input_shape;
+
+                // Get the axis where the DFT will be performed.
+                auto axis = static_cast<int>(getAttribute(ctx, "axis", 1));
+                auto rank = input_shape.dim_size();
+
+                if (!(-rank <= axis && axis < rank)) {
+                    fail_shape_inference(
+                        "axis attribute value ",
+                        axis,
+                        " is invalid for a tensor of rank ",
+                        rank);
+                }
+                
+                auto axis_idx = (axis >= 0 ? axis : axis + rank);
+                
+                // If dft_length is specified, then we should honor the shape.
+                // Set the output dimension to match the dft_length on the axis.
+                // If onesided this will be adjusted later on...
+                const TensorProto* dft_length = nullptr;
+                if (hasInputShape(ctx, 1))
```

Comment:
I don't think we need this condition and can drop it?

Reply:
dft_length is optional, it may not be supplied.

So i figured there were 3 cases.
1) Not specified -> will wont have InputShape
2) Specified, but not constant -> getInputData will fail
3) Specified, and constant -> getInputData will succeed

Is that not right?

Reply:
You are right, we must distinguish between the three cases. But I was confused by the use of `hasInputShape`. I think that is not equivalent to checking if a particular input is present or not (an input might be specified, but its shape may not be available; however, a type is always available for a specified input in a valid program). I suggest using the condition `(ctx.getNumInputs() >= 2) && (ctx.getInputType(1) != nullptr)` as a way to check if this input is specified. @jcwchen : Does this sound reasonable? Can we add a method `hasInput(i)` to `InferenceContext` that is defined to be equivalent to a check like above?

Reply:
Updated!